### PR TITLE
Various README updates

### PR DESCRIPTION
**Pull request by alsrgv**
1. Open MPI 3 specific instructions.
2. Dealing with non-routed network interfaces.
3. Keras 2.0.9 bug.
4. `-mca pml ob1` to avoid UCX issues.
5. Benchmarking instructions.

Covers questions in #70, #67, #60.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -122,33 +117,40 @@ with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
     mon_sess.run(train_op)
 ```
 
-To run on a machine with 4 GPUs:
+## Running the Horovod
 
-```bash
-$ mpirun -np 4 python train.py
-```
+Below are a couple of examples of commands to run the distributed training. See the [Running the Horovod](docs/running.md)
```
Comment (sblotner) [inline README.md]: The example commands below show how to run distributed training. See the [Running Horovod](docs/running.md)

File: README.md [hunk 2]
```diff
@@ -122,33 +117,40 @@ with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
...
-Check your MPI documentation for arguments to the `mpirun` command on your system.
-
 ## Keras
 
 Horovod supports Keras and regular TensorFlow in similar ways.
 
 See full training [simple](examples/keras_mnist.py) and [advanced](examples/keras_mnist_advanced.py) examples.
 
+**Note**: Keras 2.0.9 has a [known issue](https://github.com/fchollet/keras/issues/8353) that makes each worker allocate
+all GPUs on the server, instead of the GPU assigned by the *local rank*. If you have multiple GPUs per server, upgrade
```
Comment (sblotner) [inline README.md]: of the GPU --> of GPUs

File: docs/benchmarks.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        -mca pml ob1 -mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,65536,32 \
+        -H server1:4,server2:4,server3:4,server4:4 \
+        \
+        python scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
+            --model resnet101 \
+            --batch_size 64 \
+            --variable_update horovod
+    ```
+
+4. At the end of the run, you will see number of images processed per second:
```
Comment (sblotner) [inline docs/benchmarks.md]: see number --> see the number

File: docs/benchmarks.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+```
+total images/sec: 1656.82
+```
+
+### Real data benchmarks
+
+The benchmark instructions above are for the synthetic data benchmark.
+
+In order to run the benchmark on a real data, you need to download [ImageNet dataset](http://image-net.org/download-images)
```
Comment (sblotner) [inline docs/benchmarks.md]: In order to --> To
download ImageNet --> download the ImageNet

File: docs/benchmarks.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    python scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
+        --model resnet101 \
+        --batch_size 64 \
+        --variable_update horovod \
+        --data_dir /path/to/imagenet/tfrecords \
+        --data_name imagenet \
+        --num_batches=2000
+```
+
+2. If you have a network with RoCE / InfiniBand (recommended):
```
Comment (sblotner) [inline docs/benchmarks.md]: If this is recommended, put this option first

File: docs/running.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
+## Running the Horovod
```
Comment (sblotner) [inline docs/running.md]: remove "the"

File: docs/running.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+Examples below are for Open MPI. Check your MPI documentation for arguments to the `mpirun` command on your system.
```
Comment (sblotner) [inline docs/running.md]: Examples below --> The examples below

File: docs/running.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+Typically, one GPU will be allocated per process, so if a server has 4 GPUs you would run 4 processes. In Open MPI,
```
Comment (sblotner) [inline docs/running.md]: remove comma after "Typically" 
Add a comma after 4 GPUs

File: docs/running.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+number of processes is specified with `-np` flag.
```
Comment (sblotner) [inline docs/running.md]: number --> the number
with `-np` --> with the `-np`

File: docs/running.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+Starting with the Open MPI 3, it's important to add `-bind-to none` and `-oversubscribe` arguments. `-bind-to none`
```
Comment (sblotner) [inline docs/running.md]: to add --> to add the

File: docs/running.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+specifies Open MPI to not bind a training process to a single CPU core (which would hurt performance). `-oversubscribe`
+allows to run multiple training processes on a single core.
```
Comment (sblotner) [inline docs/running.md]: allows --> enables you to

File: docs/running.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+The `-x` option allows to specify (`-x NCCL_DEBUG=INFO`) or copy (`-x LD_LIBRARY_PATH`) an environment variable to all
```
Comment (sblotner) [inline docs/running.md]: With the `-x` option you can specify (`-x NCCL_DEBUG=INFO`) or copy (`-x LD_LIBRARY_PATH`) an environment variable to all

File: docs/running.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+```bash
+$ mpirun -np 16 \
+    -bind-to none -oversubscribe \
+    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH \
+    -H server1:4,server2:4,server3:4,server4:4 \
+    python train.py
+```
+
+3. If you have RoCE or InfiniBand, we found the following `pml` and `btl_openib_receive_queues` parameters to help
```
Comment (sblotner) [inline docs/running.md]: we found the following --> we found that the 
parameters to help --> parameters improve

File: docs/running.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    -mca pml ob1 -mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,65536,32 \
+    -H server1:4,server2:4,server3:4,server4:4 \
+    python train.py
+```
+
+### Hangs due to non-routed network interfaces
+
+Having network interfaces that are not routed can cause Open MPI to hang. An example of such interface is `docker0`.
+
+If you see non-routed interfaces (like `docker0`) in the output of `ifconfig`, you should tell Open MPI to not used them
```
Comment (sblotner) [inline docs/running.md]: used --> use

File: README.md [hunk 15]
```diff
@@ -122,33 +117,40 @@ with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
...
 ```
 
-To run on a machine with 4 GPUs:
+## Running the Horovod
 
-```bash
-$ mpirun -np 4 python train.py
-```
+Below are a couple of examples of commands to run the distributed training. See the [Running the Horovod](docs/running.md)
+page for more instructions, including RoCE/InfiniBand tweaks and dealing with hangs.
```
Comment (sblotner) [inline README.md]: dealing with hangs --> tips for dealing with hangs.

File: README.md (hunk 2 above)
Comment (alsrgv) [reply README.md]: It's actually a single GPU.

**Pull request merged**